src/adapter.rs

```diff
@@ -0,0 +1,461 @@
+       self.0.is_discovering()
+    }
+
+
+    fn create_discovery_session(&self) -> Result<BluetoothDiscoverySession, Box<Error>> {
+        BluetoothDiscoverySession::create_session(self.clone())
+    }
+
+    fn get_uuids(&self) -> Result<Vec<String>, Box<Error>> {
+        self.0.get_uuids()
+    }
+
+
+    fn get_vendor_id_source(&self) -> Result<String, Box<Error>> {
+        self.0.get_vendor_id_source()
+    }
+
+    fn get_vendor_id(&self) -> Result<u32, Box<Error>> {
+        self.0.get_vendor_id()
+    }
+
+    fn get_product_id(&self) -> Result<u32, Box<Error>> {
+        self.0.get_product_id()
+    }
+
+    fn get_device_id(&self) -> Result<u32, Box<Error>> {
+        self.0.get_device_id()
+    }
+
+    fn get_modalias(&self) -> Result<(String, u32, u32, u32), Box<Error>> {
+        self.0.get_modalias()
+    }
+}   
+
+pub struct Android(Arc<BluetoothAdapterAndroid>);
+
+impl BluetoothAdapter for Android{
+
+    fn new() -> Result<Box<BluetoothAdapter>, Box<Error>> {
+        let blurdroid_adapter = try!(BluetoothAdapter::get_adapter());
+        Ok(BluetoothAdapter::Android(Arc::new(blurdroid_adapter)))
+    }
+
+    fn get_id(&self) -> String{
+        self.0.get_id()
+    }
+
+    fn get_devices(&self) -> Result<Vec<BluetoothDevice>, Box<Error>>{
+        let device_list = try!(self.0.get_device_list());
+        Ok(device_list.into_iter().map(|device| BluetoothDevice::create_device(self.clone(), device)).collect())
+    }
+
+    fn get_device(&self, address: String) -> Result<Option<BluetoothDevice>, Box<Error>> {
+        let devices = try!(self.get_devices());
+        for device in devices {
+            if try!(device.get_address()) == address {
+                return Ok(Some(device));
+            }
+        }
+        Ok(None)
+    }
+
+    fn get_address(&self) -> Result<String, Box<Error>> {
+        self.0.get_address()
+    }
+
+    fn get_name(&self) -> Result<String, Box<Error>> {
+        self.0.get_name()
+    }
+
+
+    fn get_alias(&self) -> Result<String, Box<Error>> {
+        self.0.get_alias()
+    }
+
+
+    fn get_class(&self) -> Result<u32, Box<Error>> {
+        self.0.get_class()
+    }
+
+    fn is_powered(&self) -> Result<bool, Box<Error>> {
+        self.0.is_powered()
+    }
+
+
+    fn is_discoverable(&self) -> Result<bool, Box<Error>> {
+       self.0.is_discoverable()
+    }
+
+
+    fn is_pairable(&self) -> Result<bool, Box<Error>> {
+       self.0.is_pairable()
+    }
+
+
+    fn get_pairable_timeout(&self) -> Result<u32, Box<Error>> {
+       self.0.get_pairable_timeout()
+    }
+
+    fn get_discoverable_timeout(&self) -> Result<u32, Box<Error>> {
+       self.0.get_discoverable_timeout()
+    }
+
+    fn is_discovering(&self) -> Result<bool, Box<Error>> {
+       self.0.is_discovering()
+    }
+
+
+    fn create_discovery_session(&self) -> Result<BluetoothDiscoverySession, Box<Error>> {
+        BluetoothDiscoverySession::create_session(self.clone())
```

Comment:
Every API like this that current takes a `BluetoothAdapter` argument will need to be modified. We will need to inline the android branch of `BluetoothDiscoverySession::create_session` in this method instead (ie. create a `BluetoothDiscoverySessionBluez object directly here).